Firestore/core/src/firebase/firestore/remote/serializer.h

```diff
@@ -17,27 +17,109 @@
 #ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_SERIALIZER_H_
 #define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_SERIALIZER_H_
 
+#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.pb.h"
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
 namespace firebase {
 namespace firestore {
 namespace remote {
 
 /**
  * @brief Converts internal model objects to their equivalent protocol buffer
- * form.
+ * form, and protocol buffer objects to their equivalent bytes.
  *
- * Methods starting with "Encode" convert to a protocol buffer and methods
- * starting with "Decode" convert from a protocol buffer.
+ * Methods starting with "Encode" either convert from a model object to a
+ * protocol buffer or from a protocol buffer to bytes, and methods starting with
+ * "Decode" either convert from a protocol buffer to a model object or from
+ * bytes to a protocol buffer.
  *
  */
 // TODO(rsgowman): Original docs also has this: "Throws an exception if a
 // protocol buffer is missing a critical field or has a value we can't
 // interpret." Adjust for C++.
 class Serializer {
  public:
-  // TODO(rsgowman): Adjust ctor to accept a DatabaseID... once that exists.
-  Serializer(/*DatabaseID databaseId*/);
+  /**
+   * @brief Wraps (nanopb) google_firestore_v1beta1_Value with type information.
+   */
+  struct ValueWithType {
+    firebase::firestore::model::FieldValue::Type type;
+    google_firestore_v1beta1_Value value;
+  };
+
+  explicit Serializer(firebase::firestore::model::DatabaseId database_id)
```

Comment:
Should accept pointer (and not owning). DatabaseIs is shared.

Reply:
Copying by value was deliberate... but also lazy on my part. (Assumptions: (a) DatabaseId contains two fairly small strings and is thus cheap to copy; (b) Only a small number of Serializer objects will be constructed, typically 1, and thus the DAtabaseId won't be copied often.)

I was being lazy, because I don't know the lifetime of DatabaseIds. In other words, can we assume DatabaseIds are long-lived? (i.e. a const& would be fine as opposed to a shared_ptr?)

Reply:
Just use a non-owning raw pointer.

FirestoreClient owns the lifetime of the entire constellation of objects here including DatabaseIds and Serializers.